keps/sig-storage/20190530-pv-health-monitor.md

```diff
@@ -0,0 +1,539 @@
+* The disk that the volume resides on could fail.
+* There may be configuration issues with the underlying storage system that affect the volume’s health.
+* Volume could be out of capacity.
+* The disk may be degrading which affects its performance.
+
+If the volume is mounted on a pod and used by an application, the following problems could also happen.
+* There may be read/write I/O errors.
+* The file system on the volume may be corrupted.
+* Filesystem may be out of capacity.
+* Volume may be unmounted by accident outside of Kubernetes.
+
+If the CSI driver has implemented the CSI volume health function proposed in this design document, Kubernetes could communicate with the CSI driver to retrieve any errors detected by the underlying storage system. Kubernetes can store this data on the associated PVC so that user can inquire this information and decide how to handle them. For example, if the volume is out of capacity, user can request a volume expansion to get more space. Kubernetes may also use volume health information stored on the PV to automatically reconcile.
+
+There could be conditions that cannot be reported by a CSI driver. There could be network failure where Kubernetes may not be able to get response from the CSI driver. In this case, a call to the CSI driver may time out. There could be network congestion which causes slow response. One or more nodes where the volume is attached to may be down. This can be monitored and detected by the volume health controller so that user knows what has happened.
+
+## Proposal
+
+Volume monitoring is the main focus of this proposal. Reactions are not in the scope of this proposal.
+
+- If a volume provisioned by the CSI driver is deleted, we need to mark the corresponding PVC to inform users.
+
+- If mounting error occurs, we need to mark the PVC.
+
+- If other errors occur, we also need to mark the PVC.
+
+- If volume errors are detected by other components deployed by users, we need to mark the PVC.
+
+The main architecture is shown below:
+
+![pv health monitor architecture](./pv-health-monitor.png)
+
+The following areas will be the focus of this proposal at first:
+
+- The health condition checking of volumes themselves. For example, whether the volume is deleted, whether the usage is reaching the threshold, and so on.
+- Mounting conditions checking.
+- Other errors that could affect the usability of the volume.
+
+In addition, a service will be created to receive volume health condition reports from other compoments implemented and deployed by users.
+
+Three main parts are involved here in the architecture.
+
+- API changes: The Status field in the PVC will be used to mark volumes if they are unhealthy.
+- External Controller:
+  - Trigger controller RPC to check the health condition of network volumes themselves for network storage.
+  - Watch for node failure events for both network and local storage.
+  - Create HTTP(RPC) service to receive PVs health condition reports.
```

Comment:
how are those reports authenticated/authorized?

Reply:
how is this service accessed? is it a service routed to via kube-proxy, or does this assume network visibility from agents on nodes to the controller service?

Reply:
We really have not thought through about this HTTP(RPC) service.  This was brought up in a review comment, but it won't go through the CSI driver so it is an alternative approach.  I'm going to drop this from the current KEP.  In the future we can add it through a new KEP if needed.